src/core/lib/security/credentials/tls/grpc_tls_crl_provider.cc

```diff
@@ -114,5 +222,110 @@ std::shared_ptr<Crl> StaticCrlProvider::GetCrl(
   return it->second;
 }
 
+DirectoryReloaderCrlProviderImpl::~DirectoryReloaderCrlProviderImpl() {
+  // TODO(gtcooke94) do we need to worry about a race here?
+  if (refresh_handle_.has_value()) {
+    event_engine_->Cancel(refresh_handle_.value());
+  }
+}
+
+absl::StatusOr<std::shared_ptr<CrlProvider>> CreateDirectoryReloaderProvider(
+    absl::string_view directory, std::chrono::seconds refresh_duration,
+    std::function<void(absl::Status)> reload_error_callback) {
+  // TODO(gtcooke94) - testing with std::chrono
+  if (refresh_duration < std::chrono::seconds(1)) {
+    return absl::InvalidArgumentError("Refresh duration minimum is 60 seconds");
+  }
+  struct stat dir_stat;
+  if (stat(directory.data(), &dir_stat) != 0) {
+    return absl::InvalidArgumentError("The directory path is not valid.");
+  }
+  auto provider = std::make_shared<DirectoryReloaderCrlProviderImpl>(
+      directory, refresh_duration, reload_error_callback);
+  // TODO(gtcooke94) this could be slow to do at startup, but I think we want to
+  // make sure it's done before the provider is used.
+  absl::Status initial_status = provider->Update();
+  if (!initial_status.ok()) {
+    return initial_status;
+  }
+  provider->ScheduleReload();
+  return provider;
+}
+
+void DirectoryReloaderCrlProviderImpl::OnNextUpdateTimer() {
+  absl::Status status = Update();
+  if (!status.ok()) {
+    if (reload_error_callback_ != nullptr) {
+      reload_error_callback_(status);
+    }
+  }
+  ScheduleReload();
+}
+
+void DirectoryReloaderCrlProviderImpl::ScheduleReload() {
+  std::weak_ptr<DirectoryReloaderCrlProviderImpl> self = shared_from_this();
+  refresh_handle_ = event_engine_->RunAfter(refresh_duration_, [self]() {
+    ApplicationCallbackExecCtx callback_exec_ctx;
+    ExecCtx exec_ctx;
+    {
+      if (std::shared_ptr<DirectoryReloaderCrlProviderImpl> valid_ptr =
+              self.lock()) {
+        valid_ptr->OnNextUpdateTimer();
+      }
+    }
+  });
+}
+
+absl::Status DirectoryReloaderCrlProviderImpl::Update() {
```

Comment:
Rather than having to worry about other threads grabbing the lock between iterations of the for loop in this function, I would just consider holding a lock over `mu_` for the entire duration of this function call.